### DISCUSSION TITLE
Provide architecture.md

### DISCUSSION THREAD

**Discussion by nilsocket** [gildas-lormeau/SingleFile#688] [General]
Please provide some architecture document, giving overview of how this tool is built.

I want to build something similar in golang, can you give an high-level overview of how things work.

I had a initial implementation using chrome headless browser, but size is big,
for example, trying to store a edx webpage leads to > 3MiB in MHTML format.

Whereas using your extension, leads to < 500KiB.

Thank you.

**Comment by gildas-lormeau** [comment]
I agree that some documentation would be welcome. It requires some work though.

In the short term, to answer your question, the default settings of SingleFile are optimized to produce small files. To do this, it removes:
 - most of unused CSS rules and declarations by computing the cascade
 - stylesheets that aren't used for screen display
 - unused fonts
 - hidden elements and their CSS rules
 - responsive images that are not displayed

Some other optimizations also help to reduce the file size:
 - removal of optional HTML tags and quotes
 - minification of CSS and HTML
 - image deduplication by using CSS custom properties

**Reply by nilsocket** [reply]
Thanks a lot for your reply.

One last question, what's your input to process. i.e., do you take dom as input or some-other thing.
My front-end knowledge is very limited.

Any suggestion or any link's for further reading would help greatly.

Thank you.
